cpp/include/cudf/lists/detail/scatter.cuh

```diff
@@ -0,0 +1,727 @@
+
+    int32_t num_child_rows{get_num_child_rows(list_offsets, stream)};
+
+    auto string_views = rmm::device_vector<string_view>(num_child_rows);
+
+    auto populate_string_views = [d_scattered_lists =
+                                    list_vector.data().get(),  // unbound_list_view*
+                                  d_list_offsets = list_offsets.template data<int32_t>(),
+                                  d_string_views = string_views.data().get(),
+                                  source_lists,
+                                  target_lists] __device__(auto const& row_index) {
+      auto unbound_list_view    = d_scattered_lists[row_index];
+      auto actual_list_row      = unbound_list_view.bind_to_column(source_lists, target_lists);
+      auto lists_column         = actual_list_row.get_column();
+      auto lists_offsets_column = lists_column.offsets();
+      auto child_strings_column = lists_column.child();
+      auto string_offsets_column =
+        child_strings_column.child(cudf::strings_column_view::offsets_column_index);
+      auto string_chars_column =
+        child_strings_column.child(cudf::strings_column_view::chars_column_index);
+
+      auto output_start_offset =
+        d_list_offsets[row_index];  // Offset in `string_views` at which string_views are
+                                    // to be written for this list row_index.
+      auto input_list_start =
+        lists_offsets_column.template element<int32_t>(unbound_list_view.row_index());
+
+      thrust::for_each_n(
+        thrust::seq,
+        thrust::make_counting_iterator<size_type>(0),
+        actual_list_row.size(),
+        [output_start_offset,
+         d_string_views,
+         input_list_start,
+         d_string_offsets = string_offsets_column.template data<int32_t>(),
+         d_string_chars =
+           string_chars_column.template data<char>()] __device__(auto const& string_idx) {
+          // auto string_offset     = output_start_offset + string_idx;
+          auto string_start_idx = d_string_offsets[input_list_start + string_idx];
+          auto string_end_idx   = d_string_offsets[input_list_start + string_idx + 1];
+
+          d_string_views[output_start_offset + string_idx] =
+            string_view{d_string_chars + string_start_idx, string_end_idx - string_start_idx};
+        });
+    };
+
+    thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
+                       thrust::make_counting_iterator<size_type>(0),
+                       list_vector.size(),
+                       populate_string_views);
+
+    // string_views should now have been populated with source and target references.
+
+    auto string_offsets =
+      cudf::strings::detail::child_offsets_from_string_vector(string_views, mr, stream);
+    auto string_chars = cudf::strings::detail::child_chars_from_string_vector(
+      string_views, string_offsets->view().data<int32_t>(), 0, mr, stream);
```

Comment:
I see you're trying to use the factory here: https://github.com/rapidsai/cudf/blob/dd3eb62847b763a8579c62d2e7444e4f774e75d2/cpp%2Finclude%2Fcudf%2Fcolumn%2Fcolumn_factories.hpp#L489-L496

But you can use the factory that takes `string_view` here: https://github.com/rapidsai/cudf/blob/dd3eb62847b763a8579c62d2e7444e4f774e75d2/cpp%2Finclude%2Fcudf%2Fcolumn%2Fcolumn_factories.hpp#L383-L387

If it is unsuitable because you want to use `construct_child_nullmask` to keep the null mask logic separated then consider using `device_uvector` for the vector `string_views`

Reply:
You're right about my needing the explicit construction of the child null mask (via `construct_child_nullmask()`). This is unlike `cudf::scatter<string_view>()`, which banks on the null-mask being set at the top level.

I tried moving this code to use `rmm::device_uvector<string_view>`. The problem there is the following line:
```c++
    auto string_offsets =
      cudf::strings::detail::child_offsets_from_string_vector(string_views, mr, stream.value());
```
I'm afraid `cudf::strings::detail::child_offsets_from_string_vector()` expects an `rmm::device_vector<string_view>`. :/

Reply:
Hmmm, maybe that should be made more generic.

Reply:
I'm trying to sort this one out now.

Reply:
Fixed, at last.

Reply:
Nice utility improvement.